test/functional/testlib.bash

```diff
@@ -2659,6 +2694,429 @@ create_bundle() { # swupd_function
+			fi
+		else
+			# the file is a file or symlink, copy it to the files directory
+			# (do not follow links)
+			debug_msg "Copying file/symlink $recursive_path/$filename to $files_dir/$filehash..."
+			sudo cp -P "$recursive_path"/"$filename" "$files_dir"/"$filehash"
+		fi
+
+		# create the tar for each file, we need to skip validations since
+		# links will be dangling while in the files directory
+		debug_msg "Creating the tar for fullfile $files_dir/$filehash"
+		create_tar --skip-validation "$files_dir"/"$filehash"
+
+		# Add the file to the zero pack of the bundle
+		debug_msg "Adding fullfile $filehash to zero pack -> $version_dir/pack-$bundle_name-from-0.tar"
+		sudo tar --preserve-permissions -C "$files_dir" -rf "$version_dir"/pack-"$bundle_name"-from-0.tar --transform "s,^,staged/," "$(basename "$filehash")"
+
+		debug_msg "$relative_path/$filename added to bundle $bundle_name successfully"
+
+	done
+
+}
+
+create_bundle_from_content() { #swupd_function
+
+	cbfc_usage() {
+		cat <<-EOM
+		Usage:
+		    create_bundle_from_content [-L] [-t] [-e] [-n] <bundle_name> [-v] <version> -c <content_directory> <ENV_NAME>
+
+		Options:
+		    -c    The directory with the content that will be used to create the bundle
+		    -n    The name of the bundle to be created, if not specified a name will be autogenerated
+		    -v    The version for the bundle, if non selected, the bundle will be creted at the latest version found
+		    -L    The bundle will be 'installed' in the target-dir
+		    -t    The bundle will be 'installed' in the target-dir and 'tracked' in the tracking directory
+		    -e    The bundle will be marked as experimental
+		EOM
+	}
+
+	# variables required by getopts
+	local OPTIND
+	local opt
+	# user provided
+	local env_name
+	local bundle_name
+	local version
+	local content_dir
+	local track_bundle=false
+	local local_bundle=false
+	local experimental=false
+	# locally generated
+	local manifest
+	local web_dir
+	local version_dir
+	local target_dir
+	local state_dir
+
+	# If no parameters are received show help
+	if [ $# -eq 0 ]; then
+		cbfc_usage
+		return
+	fi
+	set -f  # turn off globbing
+	while getopts :n:c:v:tLe opt; do
+		case "$opt" in
+			n)	bundle_name="$OPTARG" ;;
+			v)	version="$OPTARG" ;;
+			c)	content_dir="$OPTARG" ;;
+			t)	track_bundle=true
+				local_bundle=true ;;
+			L)	local_bundle=true ;;
+			e)	experimental=true ;;
+			*)	cbfc_usage
+				return ;;
+		esac
+	done
+	set +f  # turn globbing back on
+	env_name=${*:$OPTIND:1}
+	validate_path "$env_name"
+	validate_path "$content_dir"
+
+	debug_msg "Creating bundle $bundle_name using content from $content_dir..."
+
+	# default values
+	bundle_name=${bundle_name:-$(generate_random_name test-bundle-)}
+	# shellcheck disable=SC2010
+	# SC2010: Don't use ls | grep. Use a glob.
+	# Exception: ls has sorting options that are tricky
+	# to get right with other commands.
+	version=${version:-$(ls "$env_name"/web-dir | grep -E '^[0-9]+$' | sort -rn | head -n1)}
```

Comment:
this is repeated 4 times. should be a function

Reply:
turns out we already have a function called `get_latest_version` which does exactly that, I had forgotten about it